docs/rfcs/20210111_centralized_configuration_management.md

```diff
@@ -0,0 +1,184 @@
+storing data in local `storage::log` it will use `raft-0` partition to replicate
+data. KV-store implementation will be realized as a simple FSM.
+All properties are cached in memory in a form of
+`absl::flat_hash_map<iobuf,iobuf>`. The distributed KV-store will use existing
+serialization library to serialize configuration properties.
+
+### Circular dependencies
+
+Since configuration module is going to use other modules, i.e. storage and
+cluster, to persist configuration properties it would have to depend on those
+modules. On the other hand those module have to depend on configuration module
+since both storage and cluster module access configuration properties.
+In order to overcome this issue configuration module will use generic
+template types which are going to be specified in `application.cc` this
+way we prevent circular dependencies.
+
+### Command line arguments support
+
+### Configuration REST API
+
+### Configuration library API
+
+### Configuration property definition
+
+### Configuration access interface
+
+
+## Interaction with other features
+
+Additional features have to be implemented in following parts of Vectorized 
+software stack:
+
+- RPK - we have to implement HTTP client that will interact with redpanda HTTP
+        admin API to manage configuration, it would be also good to implement
+        feature allowing extracting configuration from controller log segment
+        files. This would allow us to debug clusters during critical failure.
+
+- HTTP dashboard - implementing UI to navigate and  manage redpanda configuration.
+
+## Telemetry & Observability
+
+In order to be able to validate and read configuration while cluster is being
+offline we will add a tool to RPK to read configuration using configuration
+log segment files.
+
+Additional metrics for configuration change rate and configuration change.
+errors will be added.
+
+## Corner cases dissected by example
+
+One of the corner cases is setting the list of seed servers  and initialization
+of a new cluster. Right now seed server property may be different on each node.
+
+In order to simplify redpanda deployment it would be ideal if all of the nodes
+may accept the same list of seed servers.
+
+## Drawbacks
+
+The main drawback of proposed approach is an increased complexity of proposed
+solution comparing to simple configuration file. The increased complexity is
+however compensated with ease of management. User experience is one of the top
+priorities hence we should pay cost of increased complexity.
+
+## Rationale and Alternatives
+
+One of the alternatives would be to use external configuration store like
+Consul/Zookeeper/Etcd. This approach has one major drawback as it makes redpanda
+installation/operations more complicated.
+
+## Unresolved questions
+
+- Do we want to change node specific configurations from each node ?
+- Do we want to support authentication/authorization for configuration
+  management operation ?
```

Comment:
I have a use case for that in archival service. Right now configuration of the service contains aws secret key and access key that one should obtain using the aws console or cli client. These parameters should be stored on every node. This is bad because we're storing secret key on unlimited number of nodes in plain text. Better solution would be to store the signing key instead of secret key (the signing key is generated from secret key). In this case the secret key should be stored on one node. This node should calculate and propagate signing keys to all other nodes periodically (the signing key is only valid for 3 days, if I remember correctly). So, this could be implemented using this system if we will have an opportunity to set  node specific configuration parameters from other nodes.

We should probably also have some authorithation/authentification mechanism for config management and encrypted stoage for secrets.